docs/sdk-configuration.md

```diff
@@ -0,0 +1,676 @@
+        components.add(sampler);
+    }
+    
+    public OpenTelemetry build() {
+        OpenTelemetrySdk sdk = new OpenTelemetrySdk(tracerProvider.build(), meterProvider.build());
+        for (OpenTelemetryComponent component : components) {
+          component.setOpenTelemetry(sdk);
+        }
+    }
+}
+```
+
+A framework author will have an even easier time since most dependency injection frameworks
+natively support lazy injection.
+
+```java
+@Component
+public class MonitoringModule {
+  
+    @Bean
+    @ForSpanExporter
+    public Tracer tracer(TracerProvider tracerProvider) {
+    return tracerProvider.get("spanexporter");
+    }
+}
+
+@Component
+public class MyExporter implements SpanExporter {
+
+    private Lazy<Tracer> tracer;
+    
+    @Inject
+    public MyExporter(@ForSpanExporter Lazy<Tracer> tracer) {
+    this.tracer = tracer;
+    }
+    
+    @Override
+    public void export() {
+    tracer.get().spanBuilder("export").startSpan();
+    }
+}
+```
+
+## Immutability of OpenTelemetry configuration
+
+The above attempts to avoid allowing a built SDK to be mutated. Allowing mutation can make code
+harder to reason about (any component, even deep in business logic, could update the SDK without
+hindrance), can reduce performance (require volatile read on most operations), and produce thread
+safety issues if not well implemented. In particular, compared to the current state as of writing,
+
+- `addSpanProcessor` is not needed because we instead push the complexity of handling telemetry
+within telemetry components to those components, where the maintainers will have more domain
+knowledge. It allows this mutator method to be removed from the end-user API.
+
+- `updateTraceConfig` - instead of allowing updates at the top level, we should consider making
+`TraceConfig` an interface, and the SDK default implementation is static. It allows the above
+benefits of immutability to be in place for the common case where dynamic updates are not needed.
+Where dynamic updates are needed, it can be replaced with a mutable implementation instead of making
+the SDK configuration mutable. This keeps update methods out of the end-user APIs and will generally
+give framework developers more control by handling dynamicism themselves without the chance of
+end-users to affect it negatively.
+
+Some highly buggy code that could be enabled by mutability.
+
+```java
+class SleuthUsingService {
+  
+  @Inject
+  private OpenTelemetry openTelemetry;
+
+  public void doLogic() {
+    // My logic is important, so always sample it!
+    OpenTelemetrySdk.getTracerManagement().updateTraceConfig(config -> config.setSampler(ALWAYS_ON));
+    // This service was able to affect other services, even though Sleuth intends to
+    // "manage the SDK". Unlike the javaagent, it can't block access to SDK methods we may provide.
+    doSampledLogicWhileOtherServicesAlsoGetSampled();
+  }
+}
+```
+
+## TracerSdkManagement
```

Comment:
And now I understand why :) This section does not talk about SDK configuration at all :) But some completely separate concerns of public interfaces of the SDK. Should this section be in document at all?

Reply:
This came to mind while considering what's left after removing the mutability methods from management interface, since ideally we don't have a management interface at all. But agree it's not really related with others, happy to have it just in commit history.